core/src/serve_repair.rs

```diff
@@ -700,6 +798,22 @@ impl ServeRepair {
             }
 
             let from_addr = packet.meta.socket_addr();
+            let request_sender = *request.sender();
+
+            // Charge a minimum to process request.
+            // This will prevent processing if the budget has been exhausted.
+            const GOOD_FAITH_CHARGE: u64 = 10;
+            let precharge = if matches!(&request, RepairProtocol::Pong(_)) {
+                // Pong is exempt as it does not generate a response
+                0
+            } else {
+                if !repair_budget.take(&request_sender, GOOD_FAITH_CHARGE, stats) {
+                    stats.dropped_requests_outbound_bandwidth += 1;
+                    continue;
+                }
+                GOOD_FAITH_CHARGE
+            };
```

Comment:
I don't get the motivation of this. If repair budget is exhausted we were `break`ing out of the loop.
You have removed the `break` and instead adding this. Why?

If the goal is to process the pong messages, just where the code was `break` ing we could process the remaining pong messages, and skip everything else. Wouldn't that be a lot simpler?

Reply:
This is a simple heuristic to try to prevent doing additional work if we'd fail the bandwidth check later. We currently blindly do all the work of querying blockstore and creating response packet(s) only to determine that we've exhausted bandwidth and throw out the response.

Reply:
I don't expect it to be significant since it is only the very last generated response.
But if we don't want to waste the computation which is already done, I think a much simpler solution would be to send the very last generated response even if it goes above the budget, then `break` ing out; i.e. let it overshoot the budget a little bit.